File: localdb.cc
/* Cumulus: Efficient Filesystem Backup to the Cloud
 * Copyright (C) 2007-2008 The Cumulus Developers
 * See the AUTHORS file for a list of contributors.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

/* When creating backup snapshots, maintain a local database of data blocks and
 * checksums, in addition to the data contents (which may be stored remotely).
 * This database is consulted when attempting to build incremental snapshots,
 * as it says which objects can be reused.
 *
 * The database is implemented as an SQLite3 database, but this implementation
 * detail is kept internal to this file, so that the storage format may be
 * changed later. */

#include <assert.h>
#include <stdio.h>
#include <string.h>
#include <sqlite3.h>

#include <algorithm>
#include <string>

#include "localdb.h"
#include "store.h"
#include "util.h"

using std::max;
using std::min;
using std::set;
using std::string;

static const int SCHEMA_MAJOR = 0;
static const int SCHEMA_MINOR = 11;

/* Helper function to prepare a statement for execution in the current
 * database. */
sqlite3_stmt *LocalDb::Prepare(const char *sql)
{
    sqlite3_stmt *stmt;
    int rc;
    const char *tail;

    rc = sqlite3_prepare_v2(db, sql, strlen(sql), &stmt, &tail);
    if (rc != SQLITE_OK) {
        ReportError(rc);
        fatal(string("Error preparing statement: ") + sql);
    }

    return stmt;
}

void LocalDb::ReportError(int rc)
{
    fprintf(stderr, "Result code: %d\n", rc);
    fprintf(stderr, "Error message: %s\n", sqlite3_errmsg(db));
}

void LocalDb::Open(const char *path, const char *snapshot_name,
                   const char *snapshot_scheme, double intent)
{
    int rc;

    rc = sqlite3_open(path, &db);
    if (rc) {
        fprintf(stderr, "Can't open database: %s\n", sqlite3_errmsg(db));
        sqlite3_close(db);
        fatal("Error opening local database");
    }

    rc = sqlite3_exec(db, "begin", NULL, NULL, NULL);
    if (rc) {
        fprintf(stderr, "Can't open database: %s\n", sqlite3_errmsg(db));
        sqlite3_close(db);
        fatal("Error starting transaction");
    }

    sqlite3_extended_result_codes(db, 1);

    /* Check that the local database format is the correct version; if not,
     * report an error. */
    sqlite3_stmt *stmt = Prepare("select major, minor from schema_version");

    rc = sqlite3_step(stmt);
    if (rc != SQLITE_ROW) {
        fatal("Unable to read local database version from database");
    } else if (rc == SQLITE_ROW) {
        int major = sqlite3_column_int(stmt, 0);
        int minor = sqlite3_column_int(stmt, 1);
        if (major != SCHEMA_MAJOR || minor != SCHEMA_MINOR) {
            fprintf(stderr,
                    "Local database does not have required schema version!\n"
                    "  expected: %d.%d, found: %d.%d\n",
                    SCHEMA_MAJOR, SCHEMA_MINOR, major, minor);
            fatal("Unable to continue.");
        }
    }
    sqlite3_finalize(stmt);

    if (snapshot_scheme == NULL)
        snapshot_scheme = "";

    /* Insert this snapshot into the database, and determine the integer key
     * which will be used to identify it. */
    stmt = Prepare("insert into snapshots(name, scheme, timestamp, intent) "
                   "values (?, ?, julianday('now'), ?)");
    sqlite3_bind_text(stmt, 1, snapshot_name, strlen(snapshot_name),
                      SQLITE_TRANSIENT);
    sqlite3_bind_text(stmt, 2, snapshot_scheme, strlen(snapshot_scheme),
                      SQLITE_TRANSIENT);
    sqlite3_bind_double(stmt, 3, intent);

    rc = sqlite3_step(stmt);
    if (rc != SQLITE_DONE) {
        ReportError(rc);
        sqlite3_close(db);
        fatal("Database execution error!");
    }

    snapshotid = sqlite3_last_insert_rowid(db);
    sqlite3_finalize(stmt);
    if (snapshotid == 0) {
        ReportError(rc);
        sqlite3_close(db);
        fatal("Find snapshot id");
    }

    /* Create a temporary table which will be used to keep track of the objects
     * used by this snapshot.  When the database is closed, we will summarize
     * the results of this table into segments_used. */
    rc = sqlite3_exec(db,
                      "create temporary table snapshot_refs ("
                      "    segmentid integer not null,"
                      "    object text not null,"
                      "    size integer not null"
                      ")", NULL, NULL, NULL);
    if (rc != SQLITE_OK) {
        ReportError(rc);
        sqlite3_close(db);
        fatal("Database initialization");
    }
    rc = sqlite3_exec(db,
                      "create unique index snapshot_refs_index "
                      "on snapshot_refs(segmentid, object)",
                      NULL, NULL, NULL);
    if (rc != SQLITE_OK) {
        ReportError(rc);
        sqlite3_close(db);
        fatal("Database initialization");
    }
}

void LocalDb::Close()
{
    int rc;

    /* Summarize the snapshot_refs table into segment_utilization. */
    sqlite3_stmt *stmt = Prepare(
        "insert or replace into segment_utilization "
        "select ? as snapshotid, segmentid, sum(size) "
        "from snapshot_refs group by segmentid"
    );
    sqlite3_bind_int64(stmt, 1, snapshotid);
    rc = sqlite3_step(stmt);
    if (rc != SQLITE_OK && rc != SQLITE_DONE) {
        ReportError(rc);
        sqlite3_close(db);
        fprintf(stderr, "DATABASE ERROR: Unable to create segment summary!\n");
    }
    sqlite3_finalize(stmt);

    /* Commit changes to the database and close. */
    rc = sqlite3_exec(db, "commit", NULL, NULL, NULL);
    if (rc != SQLITE_OK) {
        fprintf(stderr, "DATABASE ERROR: Can't commit database!\n");
        ReportError(rc);
    }
    sqlite3_close(db);
}

int64_t LocalDb::SegmentToId(const string &segment)
{
    int rc;
    sqlite3_stmt *stmt;
    int64_t result;

    stmt = Prepare("insert or ignore into segments(segment) values (?)");
    sqlite3_bind_text(stmt, 1, segment.c_str(), segment.size(),
                      SQLITE_TRANSIENT);
    rc = sqlite3_step(stmt);
    if (rc != SQLITE_DONE) {
        fatal("Could not execute INSERT statement!");
    }
    sqlite3_finalize(stmt);

    stmt = Prepare("select segmentid from segments where segment = ?");
    sqlite3_bind_text(stmt, 1, segment.c_str(), segment.size(),
                      SQLITE_TRANSIENT);

    rc = sqlite3_step(stmt);
    if (rc == SQLITE_DONE) {
        fatal("No segment found by id");
    } else if (rc == SQLITE_ROW) {
        result = sqlite3_column_int64(stmt, 0);
    } else {
        fatal("Error executing find segment by id query");
    }

    sqlite3_finalize(stmt);

    return result;
}

string LocalDb::IdToSegment(int64_t segmentid)
{
    int rc;
    sqlite3_stmt *stmt;
    string result;

    stmt = Prepare("select segment from segments where segmentid = ?");
    sqlite3_bind_int64(stmt, 1, segmentid);

    rc = sqlite3_step(stmt);
    if (rc == SQLITE_DONE) {
        fatal("No segment found by id");
    } else if (rc == SQLITE_ROW) {
        result = (const char *)sqlite3_column_text(stmt, 0);
    } else {
        fatal("Error executing find segment by id query");
    }

    sqlite3_finalize(stmt);

    return result;
}

void LocalDb::StoreObject(const ObjectReference& ref, double age)
{
    int rc;
    sqlite3_stmt *stmt;

    assert(ref.has_checksum());
    string checksum = ref.get_checksum();
    assert(ref.range_is_exact());
    int64_t size = ref.get_range_length();

    if (age == 0.0) {
        stmt = Prepare("insert into block_index("
                       "segmentid, object, checksum, size, timestamp) "
                       "values (?, ?, ?, ?, julianday('now'))");
    } else {
        stmt = Prepare("insert into block_index("
                       "segmentid, object, checksum, size, timestamp) "
                       "values (?, ?, ?, ?, ?)");
    }

    sqlite3_bind_int64(stmt, 1, SegmentToId(ref.get_segment()));
    string obj = ref.get_sequence();
    sqlite3_bind_text(stmt, 2, obj.c_str(), obj.size(), SQLITE_TRANSIENT);
    sqlite3_bind_text(stmt, 3, checksum.c_str(), checksum.size(),
                      SQLITE_TRANSIENT);
    sqlite3_bind_int64(stmt, 4, size);
    if (age != 0.0)
        sqlite3_bind_double(stmt, 5, age);

    rc = sqlite3_step(stmt);
    if (rc != SQLITE_DONE) {
        fprintf(stderr, "Could not execute INSERT statement!\n");
        ReportError(rc);
    }

    sqlite3_finalize(stmt);
}

ObjectReference LocalDb::FindObject(const string &checksum, int64_t size)
{
    int rc;
    sqlite3_stmt *stmt;
    ObjectReference ref;

    stmt = Prepare("select segmentid, object from block_index "
                   "where checksum = ? and size = ? and expired is null");
    sqlite3_bind_text(stmt, 1, checksum.c_str(), checksum.size(),
                      SQLITE_TRANSIENT);
    sqlite3_bind_int64(stmt, 2, size);

    rc = sqlite3_step(stmt);
    if (rc == SQLITE_DONE) {
    } else if (rc == SQLITE_ROW) {
        ref = ObjectReference(IdToSegment(sqlite3_column_int64(stmt, 0)),
                              (const char *)sqlite3_column_text(stmt, 1));
        ref.set_range(0, size, true);
    } else {
        fprintf(stderr, "Could not execute SELECT statement!\n");
        ReportError(rc);
    }

    sqlite3_finalize(stmt);

    return ref;
}

bool LocalDb::IsOldObject(const string &checksum, int64_t size, double *age,
                          int *group)
{
    int rc;
    sqlite3_stmt *stmt;
    bool found = false;

    stmt = Prepare("select segmentid, object, timestamp, expired "
                   "from block_index where checksum = ? and size = ?");
    sqlite3_bind_text(stmt, 1, checksum.c_str(), checksum.size(),
                      SQLITE_TRANSIENT);
    sqlite3_bind_int64(stmt, 2, size);

    rc = sqlite3_step(stmt);
    if (rc == SQLITE_DONE) {
        found = false;
    } else if (rc == SQLITE_ROW) {
        found = true;
        *age = sqlite3_column_double(stmt, 2);
        *group = sqlite3_column_int(stmt, 3);
    } else {
        fprintf(stderr, "Could not execute SELECT statement!\n");
        ReportError(rc);
    }

    sqlite3_finalize(stmt);

    return found;
}

/* Does this object still exist in the database (and not expired)? */
bool LocalDb::IsAvailable(const ObjectReference &ref)
{
    int rc;
    sqlite3_stmt *stmt;
    bool found = false;

    // Special objects (such as the zero object) aren't stored in segments, and
    // so are always available.
    if (!ref.is_normal())
        return true;

    stmt = Prepare("select count(*) from block_index "
                   "where segmentid = ? and object = ? and expired is null");
    sqlite3_bind_int64(stmt, 1, SegmentToId(ref.get_segment()));
    sqlite3_bind_text(stmt, 2, ref.get_sequence().c_str(),
                      ref.get_sequence().size(), SQLITE_TRANSIENT);

    rc = sqlite3_step(stmt);
    if (rc == SQLITE_DONE) {
        found = false;
    } else if (rc == SQLITE_ROW) {
        if (sqlite3_column_int(stmt, 0) > 0)
            found = true;
    } else {
        fprintf(stderr, "Could not execute SELECT statement!\n");
        ReportError(rc);
    }

    sqlite3_finalize(stmt);

    return found;
}

set<string> LocalDb::GetUsedSegments()
{
    int rc;
    sqlite3_stmt *stmt;
    set<string> result;

    stmt = Prepare("select segment from segments "
                   "where segmentid in (select segmentid from snapshot_refs)");

    while (true) {
        rc = sqlite3_step(stmt);
        if (rc == SQLITE_ROW) {
            const char *segment
                = reinterpret_cast<const char *>(sqlite3_column_text(stmt, 0));
            result.insert(segment);
        } else if (rc == SQLITE_DONE) {
            break;
        } else {
            ReportError(rc);
        }
    }

    sqlite3_finalize(stmt);

    return result;
}

void LocalDb::UseObject(const ObjectReference& ref)
{
    int rc;
    sqlite3_stmt *stmt;

    if (!ref.is_normal())
        return;

    int64_t old_size = 0;
    stmt = Prepare("select size from snapshot_refs "
                   "where segmentid = ? and object = ?");
    sqlite3_bind_int64(stmt, 1, SegmentToId(ref.get_segment()));
    string obj = ref.get_sequence();
    sqlite3_bind_text(stmt, 2, obj.c_str(), obj.size(), SQLITE_TRANSIENT);
    rc = sqlite3_step(stmt);
    if (rc == SQLITE_ROW) {
        old_size = sqlite3_column_int64(stmt, 0);
    }
    sqlite3_finalize(stmt);

    // Attempt to determine the underlying size of the object.  This may
    // require a database lookup if the length is not encoded into the object
    // reference already.
    int64_t object_size = 0;
    if (ref.range_is_exact()) {
        object_size = ref.get_range_length();
    } else {
        stmt = Prepare("select size from block_index "
                       "where segmentid = ? and object = ?");
        sqlite3_bind_int64(stmt, 1, SegmentToId(ref.get_segment()));
        obj = ref.get_sequence();
        sqlite3_bind_text(stmt, 2, obj.c_str(), obj.size(), SQLITE_TRANSIENT);
        rc = sqlite3_step(stmt);
        if (rc == SQLITE_ROW) {
            object_size = sqlite3_column_int64(stmt, 0);
        } else {
            fprintf(stderr, "Warning: No block found in block_index for %s\n",
                    ref.to_string().c_str());
        }
        sqlite3_finalize(stmt);
    }

    // Possibly mark additional bytes as being referenced.  The number of bytes
    // referenced can only be increased (up to the object size).  The bytes
    // referenced will be set to the object size only if the entire object is
    // referenced at once: a series of partial ranges that add up to the total
    // size will have a reference size capped at just less than the full object
    // size (we can't tell if some bytes were referenced multiple times, and
    // thus we conservatively assume some bytes might still be unreferenced).
    int64_t new_refs = old_size;
    if (ref.has_range()) {
        new_refs = ref.get_range_length();
    } else {
        new_refs = object_size;
    }
    int64_t new_size = old_size + new_refs;
    if (old_size < object_size && new_refs < object_size)
        new_size = min(new_size, object_size - 1);
    new_size = min(object_size, new_size);
    new_size = max(new_size, (int64_t)0);

    if (new_size != old_size) {
        stmt = Prepare("insert or replace "
                       "into snapshot_refs(segmentid, object, size) "
                       "values (?, ?, ?)");
        sqlite3_bind_int64(stmt, 1, SegmentToId(ref.get_segment()));
        obj = ref.get_sequence();
        sqlite3_bind_text(stmt, 2, obj.c_str(), obj.size(), SQLITE_TRANSIENT);
        sqlite3_bind_int64(stmt, 3, new_size);

        rc = sqlite3_step(stmt);
        if (rc != SQLITE_DONE) {
            fprintf(stderr, "Could not execute INSERT statement!\n");
            ReportError(rc);
        }

        sqlite3_finalize(stmt);
    }
}

void LocalDb::SetSegmentMetadata(const std::string &segment,
                                 const std::string &path,
                                 const std::string &checksum,
                                 const std::string &type,
                                 int data_size, int disk_size)
{
    int rc;
    sqlite3_stmt *stmt;

    stmt = Prepare("update segments set path = ?, checksum = ?, "
                   "type = ?, data_size = ?, disk_size = ?, "
                   "mtime = coalesce(mtime, julianday('now')) "
                   "where segmentid = ?");
    sqlite3_bind_text(stmt, 1, path.c_str(), path.size(),
                      SQLITE_TRANSIENT);
    sqlite3_bind_text(stmt, 2, checksum.c_str(), checksum.size(),
                      SQLITE_TRANSIENT);
    sqlite3_bind_text(stmt, 3, type.c_str(), type.size(),
                      SQLITE_TRANSIENT);
    sqlite3_bind_int64(stmt, 4, data_size);
    sqlite3_bind_int64(stmt, 5, disk_size);
    sqlite3_bind_int64(stmt, 6, SegmentToId(segment));

    rc = sqlite3_step(stmt);
    if (rc != SQLITE_DONE) {
        fprintf(stderr, "Could not update segment checksum in database!\n");
        ReportError(rc);
    }

    sqlite3_finalize(stmt);
}

bool LocalDb::GetSegmentMetadata(const string &segment,
                                 string *seg_path,
                                 string *seg_checksum)
{
    int rc;
    sqlite3_stmt *stmt;
    ObjectReference ref;
    int found = false;

    stmt = Prepare("select path, checksum from segments where segment = ?");
    sqlite3_bind_text(stmt, 1, segment.c_str(), segment.size(),
                      SQLITE_TRANSIENT);

    rc = sqlite3_step(stmt);
    if (rc == SQLITE_DONE) {
    } else if (rc == SQLITE_ROW) {
        found = true;
        const char *val;

        val = (const char *)sqlite3_column_text(stmt, 0);
        if (val == NULL)
            found = false;
        else
            *seg_path = val;

        val = (const char *)sqlite3_column_text(stmt, 1);
        if (val == NULL)
            found = false;
        else
            *seg_checksum = val;
    } else {
        fprintf(stderr, "Could not execute SELECT statement!\n");
        ReportError(rc);
    }

    sqlite3_finalize(stmt);

    return found;
}

/* Look up and return the packed representation of the subblock chunk
 * signatures.  Returns true if signatures were found for the specified object,
 * and if so sets *buf to point at a buffer of memory (allocated with malloc;
 * the caller should free it), and *len to the length of the buffer. */
bool LocalDb::LoadChunkSignatures(ObjectReference ref,
                                  void **buf, size_t *len,
                                  string *algorithm)
{
    int rc;
    sqlite3_stmt *stmt;
    int found = false;

    stmt = Prepare("select signatures, algorithm from subblock_signatures "
                   "where blockid = (select blockid from block_index "
                   "                 where segmentid = ? and object = ?)");
    sqlite3_bind_int64(stmt, 1, SegmentToId(ref.get_segment()));
    string obj = ref.get_sequence();
    sqlite3_bind_text(stmt, 2, obj.c_str(), obj.size(), SQLITE_TRANSIENT);

    rc = sqlite3_step(stmt);
    if (rc == SQLITE_DONE) {
    } else if (rc == SQLITE_ROW) {
        const void *data = sqlite3_column_blob(stmt, 0);
        *len = sqlite3_column_bytes(stmt, 0);

        if (*len > 0) {
            *buf = malloc(*len);
            if (*buf != NULL) {
                memcpy(*buf, data, *len);
                *algorithm = (const char *)sqlite3_column_text(stmt, 1);
                found = true;
            }
        }
    } else {
        fprintf(stderr, "Could not execute SELECT statement!\n");
        ReportError(rc);
    }

    sqlite3_finalize(stmt);

    return found;
}

/* Store the subblock chunk signatures for a specified object.  The object
 * itself must have already been indexed in the database. */
void LocalDb::StoreChunkSignatures(ObjectReference ref,
                                   const void *buf, size_t len,
                                   const string& algorithm)
{
    int rc;
    sqlite3_stmt *stmt;

    stmt = Prepare("select blockid from block_index "
                   "where segmentid = ? and object = ?");
    sqlite3_bind_int64(stmt, 1, SegmentToId(ref.get_segment()));
    string obj = ref.get_sequence();
    sqlite3_bind_text(stmt, 2, obj.c_str(), obj.size(), SQLITE_TRANSIENT);

    rc = sqlite3_step(stmt);
    if (rc != SQLITE_ROW) {
        fprintf(stderr,
                "Could not determine blockid in StoreChunkSignatures!\n");
        ReportError(rc);
        fatal("Error getting blockid");
    }
    int64_t blockid = sqlite3_column_int64(stmt, 0);
    sqlite3_finalize(stmt);

    stmt = Prepare("insert or replace "
                   "into subblock_signatures(blockid, algorithm, signatures) "
                   "values (?, ?, ?)");
    sqlite3_bind_int64(stmt, 1, blockid);
    sqlite3_bind_text(stmt, 2, algorithm.c_str(), algorithm.size(),
                      SQLITE_TRANSIENT);
    sqlite3_bind_blob(stmt, 3, buf, len, SQLITE_TRANSIENT);

    rc = sqlite3_step(stmt);
    if (rc != SQLITE_DONE) {
        fprintf(stderr, "Could not insert sub-block checksums!\n");
        ReportError(rc);
    }

    sqlite3_finalize(stmt);
}

File: python/cumulus/localdb.py
# Cumulus: Efficient Filesystem Backup to the Cloud
# Copyright (C) 2014 The Cumulus Developers
# See the AUTHORS file for a list of contributors.
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

"""Python interface to the Cumulus local database of snapshot data.

Includes interfaces for iterating over data and implementing database and
segment cleaning.
"""

from __future__ import division, print_function, unicode_literals

import collections
import datetime
import os
import sqlite3

SnapshotInfo = collections.namedtuple(
    "SnapshotInfo",
    ["id", "scheme", "name", "timestamp"])

SegmentInfo = collections.namedtuple(
    "SegmentInfo",
    ["id", "name", "timestamp", "data_size", "disk_size", "type"])

SegmentStatistics = collections.namedtuple(
    "SegmentStatistics",
    ["id", "name", "timestamp", "data_size", "disk_size", "type",
     "bytes_referenced", "utilization"])

class Database:
    """Access to the local database of snapshot contents and object checksums.

    The local database is consulted when creating a snapshot to determine what
    data can be re-used from old snapshots.  Segment cleaning is performed by
    manipulating the data in the local database; the local database also
    includes enough data to guide the segment cleaning process.
    """

    def __init__(self, path, dbname="localdb.sqlite"):
        self.db_connection = sqlite3.connect(
            os.path.join(path, dbname),
            detect_types=sqlite3.PARSE_COLNAMES)

    @staticmethod
    def _get_id(item):
        """Fetch the id of a database object.

        If the input is an integer return it directly, otherwise try to fetch
        the .id field on it.
        """
        if isinstance(item, int):
            return item
        else:
            return item.id

    # Low-level database access.  Use these methods when there isn't a
    # higher-level interface available.  Exception: do, however, remember to
    # use the commit() method after making changes to make sure they are
    # actually saved, even when going through higher-level interfaces.
    def commit(self):
        "Commit any pending changes to the local database."
        self.db_connection.commit()

    def rollback(self):
        "Roll back any pending changes to the local database."
        self.db_connection.rollback()

    def cursor(self):
        "Return a DB-API cursor for directly accessing the local database."
        return self.db_connection.cursor()

    def get_snapshots(self):
        """Returns information about all snapshots in the local database.

        The returned value is a dictionary mapping scheme names to lists of
        snapshots for that scheme.  Each list entry is a SnapshotInfo instance.
        """
        cur = self.cursor()
        cur.execute("""select snapshotid, scheme, name,
                              datetime(timestamp) as "timestamp [timestamp]"
                       from snapshots order by scheme, name""")
        snapshots = {}
        for row in cur.fetchall():
            info = SnapshotInfo(*row)
            snapshots.setdefault(info.scheme, []).append(info)
        return snapshots

    def delete_snapshot(self, snapshot):
        """Remove the specified snapshot from the database.

        Returns a boolean indicating whether the snapshot was deleted.

        Warning: This does not garbage collect all dependent data in the
        database, so it must be followed by a call to garbage_collect() to make
        the database consistent.
        """
        cur = self.cursor()
        cur.execute("delete from snapshots where snapshotid = ?",
                    (self._get_id(snapshot),))
        return cur.rowcount > 0

    def garbage_collect(self):
        """Garbage-collect unreachable segment and object data.

        Remove all segments and checksums which is not reachable from the
        current set of snapshots stored in the local database.
        """
        cur = self.cursor()

        # Delete entries in the segment_utilization table which are for
        # non-existent snapshots.
        cur.execute("""delete from segment_utilization
                       where snapshotid not in
                           (select snapshotid from snapshots)""")

        # Delete segments not referenced by any current snapshots.
        cur.execute("""delete from segments where segmentid not in
                           (select segmentid from segment_utilization)""")

        # Delete dangling objects in the block_index table.
        cur.execute("""delete from block_index
                       where segmentid not in
                           (select segmentid from segments)""")

        # Remove sub-block signatures for deleted objects.
        cur.execute("""delete from subblock_signatures
                       where blockid not in
                           (select blockid from block_index)""")

    def get_segment_info(self):
        """Retrieve statistics about segments for cleaning decisions."""
        cur = self.cursor()
        cur.execute("""select segmentid, segment,
                              datetime(timestamp) as "timestamp [timestamp]",
                              data_size, disk_size, type
                       from segments""")
        return dict((x[0], SegmentInfo(*x)) for x in cur.fetchall())

    def get_segment_utilizations(self, snapshots):
        """Computes estimates for the data referenced in each segment.

        Computes a lower bound of the amount of data that is referenced in
        segments by the specified set of snapshots.
        """
        cur = self.cursor()
        segment_info = self.get_segment_info()
        snapshots = [self._get_id(s) for s in snapshots]

        query = """select segmentid, max(bytes_referenced)
                   from segment_utilization where snapshotid in (%s)
                   group by segmentid""" % (",".join(["?"] * len(snapshots)))
        segments = {}
        for row in cur.execute(query, snapshots):
            info = segment_info[row[0]]
            segments[row[0]] = SegmentStatistics(
                bytes_referenced=row[1],
                utilization=row[1]/info.data_size,
                **info._asdict())
        return segments

    def mark_segment_expired(self, segment):
        """Mark a segment for cleaning in the local database.

        The segment parameter should be either a SegmentInfo object or an
        integer segment id.  Objects in the given segment will be marked as
        expired, which means that any future snapshots that would re-use those
        objects will instead write out a new copy of the object, and thus no
        future snapshots will depend upon the given segment.
        """
        cur = self.cursor()
        cur.execute("update block_index set expired = 0 where segmentid = ?",
                    (self._get_id(segment),))

def run_cleaner(database):
    # Find the most recent snapshot for each backup scheme, then delete all
    # older snapshots from the database.
    kept_snapshots = []
    for snapshots in database.get_snapshots().values():
        snapshots = sorted(snapshots, key=lambda s: s.id)
        kept_snapshots.append(snapshots[-1])
        for s in snapshots[:-1]:
            print("Deleting snapshot", s)
            database.delete_snapshot(s)
    print("Keeping snapshots:", kept_snapshots)
    database.garbage_collect()

    # TODO: Look at adding more complex policies later, perhaps under user
    # control.  Cleaning policies to consider (which can be combined):
    #   - clean older than specified age
    #   - do not clean younger than specified age
    #   - clean below utilization threshold
    #   - do not allow data from previously-unreferenced segments
    #   - gradual: clean a fraction of segments which match a rule
    #   - minimum segment size (disk or data)
    #   - benefit calculation?

    for segment in database.get_segment_utilizations(kept_snapshots).values():
        if segment.utilization < 0.4:
            print("Clean segment:", segment)
            database.mark_segment_expired(segment)

    database.commit()

if __name__ == "__main__":
    # Demo usage: runs a cleaner on the database located in the current
    # directory.  This should be removed later when cleaning is properly hooked
    # up to the main tool.
    run_cleaner(Database("."))
